src/rust/src/lib.rs

```diff
@@ -0,0 +1,31 @@
+#![crate_type = "staticlib"]
+#![allow(non_snake_case)]
+
+use cxx::{CxxVector,CxxString};
+use std::pin::Pin;
+
+// This module (mod bridge) is the signature for this crate that's exported to C++
+#[cxx::bridge(namespace="stellar::bridge")]
+mod bridge {
+    extern "Rust" {
+        fn to_base64(b: &CxxVector<u8>, mut s: Pin<&mut CxxString>);
+        fn from_base64(s: &CxxString, mut b: Pin<&mut CxxVector<u8>>);
+    }
+}
+
+// This is the rust implementation of the bridge signature.
+fn to_base64(b: &CxxVector<u8>, mut s: Pin<&mut CxxString>)
+{
+    s.as_mut().clear();
+    s.push_str(base64::encode(b.as_slice()).as_str());
+}
+
+fn from_base64(s: &CxxString, mut b: Pin<&mut CxxVector<u8>>)
+{
+    assert!(b.len() == 0);
+    let ss = s.to_str().expect("non-str base64 string");
```

Comment:
I guess related to this, what's the interop story when it comes to errors and exceptions? Is it just panic all around (with the appropriate crash handler being called to log & exit)?

Should the bridge function exposed to c++ be marked "noexcept" (and end up calling "std::terminate" on panic somehow) and do we want bridge functions to convert errors into exceptions (for the non panic case)?

What about the other way around? Are we going to call C++ from Rust? If so, we should also enforce that whatever is called is "noexcept" on the C++ side

Reply:
@jonjove  I've pushed an update that makes it "do what the C++ does" which is just silently eat malformed bytes in the base64 input. woo. we might want to make this actually strict at some point.

@MonsieurNicolas the default behaviour is to [map Rust's `Result<T>` type to and from C++ exceptions](https://cxx.rs/binding/result.html) -- both directions, so it'll catch C++ exceptions and turn them _into_ Rust `Result<T>` values too -- with a little bit of customization available. They're going to continue to work on this part of the bridge but since we do use exceptions it's probably close to ok for our needs now.

I just, in this case, didn't return a Result because the existing C++ code had no error-handling paths. I assumed this was because we sanitized its input, but it's actually because it's error tolerant (perhaps unwisely, but ..)

Reply:
Opened #3374